File: src/core/pathfinding.js
class MinHeap {
  constructor() {
    this.items = [];
  }

  push(node) {
    this.items.push(node);
    let index = this.items.length - 1;
    while (index > 0) {
      const parent = (index - 1) >> 1;
      if (this.items[parent].score <= node.score) break;
      this.items[index] = this.items[parent];
      index = parent;
    }
    this.items[index] = node;
  }

  pop() {
    const root = this.items[0];
    const tail = this.items.pop();
    if (this.items.length && tail) {
      let index = 0;
      while (true) {
        let child = index * 2 + 1;
        if (child >= this.items.length) break;
        if (child + 1 < this.items.length && this.items[child + 1].score < this.items[child].score) child += 1;
        if (this.items[child].score >= tail.score) break;
        this.items[index] = this.items[child];
        index = child;
      }
      this.items[index] = tail;
    }
    return root;
  }

  get length() {
    return this.items.length;
  }
}

const DIRECTIONS = [[1, 0], [-1, 0], [0, 1], [0, -1]];

export function findTerrainPath(width, height, terrain, start, goal, variation) {
  const total = width * height;
  const gScore = new Float64Array(total);
  gScore.fill(Number.POSITIVE_INFINITY);
  const previous = new Int32Array(total);
  previous.fill(-1);
  const closed = new Uint8Array(total);
  const startIndex = start.y * width + start.x;
  const goalIndex = goal.y * width + goal.x;
  const heap = new MinHeap();
  gScore[startIndex] = 0;
  heap.push({ index: startIndex, score: 0 });

  while (heap.length) {
    const current = heap.pop();
    if (closed[current.index]) continue;
    if (current.index === goalIndex) break;
    closed[current.index] = 1;
    const x = current.index % width;
    const y = Math.floor(current.index / width);
    for (const [dx, dy] of DIRECTIONS) {
      const nx = x + dx;
      const ny = y + dy;
      if (nx < 1 || ny < 1 || nx >= width - 1 || ny >= height - 1) continue;
      const nextIndex = ny * width + nx;
      if (closed[nextIndex]) continue;
      const tile = terrain[nextIndex];
      const terrainCost = tile === "water" ? 11 : tile === "forest" ? 2.7 : 1;
      const organicCost = variation(nx, ny) * 0.85;
      const tentative = gScore[current.index] + terrainCost + organicCost;
      if (tentative >= gScore[nextIndex]) continue;
      previous[nextIndex] = current.index;
      gScore[nextIndex] = tentative;
      const heuristic = Math.abs(goal.x - nx) + Math.abs(goal.y - ny);
      heap.push({ index: nextIndex, score: tentative + heuristic });
    }
  }

  if (previous[goalIndex] === -1 && goalIndex !== startIndex) return [];
  const path = [];
  let index = goalIndex;
  while (index !== -1) {
    path.push({ x: index % width, y: Math.floor(index / width) });
    if (index === startIndex) break;
    index = previous[index];
  }
  return path.reverse();
}

File: native/Core/Pathfinding.cs
namespace Village.Core;

public readonly record struct Point(int X, int Y);

/// <summary>Stable binary heap matching the original JavaScript's tie ordering.</summary>
internal sealed class StableMinHeap<T>(Func<T, double> score)
{
    private readonly List<T> items = [];
    public int Count => items.Count;
    public void Push(T item)
    {
        items.Add(item); int index = items.Count - 1;
        while (index > 0) { int parent = (index - 1) >> 1; if (score(items[parent]) <= score(item)) break; items[index] = items[parent]; index = parent; }
        items[index] = item;
    }
    public T Pop()
    {
        T first = items[0], last = items[^1]; items.RemoveAt(items.Count - 1);
        if (items.Count > 0)
        {
            int index = 0;
            while (true)
            {
                int left = index * 2 + 1, right = left + 1; if (left >= items.Count) break;
                int child = right < items.Count && score(items[right]) < score(items[left]) ? right : left;
                if (score(items[child]) >= score(last)) break; items[index] = items[child]; index = child;
            }
            items[index] = last;
        }
        return first;
    }
}

public static class Pathfinding
{
    private record State(int Index, int Direction, int WaterRun, int Key, double Cost, double Score);
    public static List<Point> FindPath(int width, int height, Point start, Point goal,
        Func<int,int,int,int,double> costAt, Func<int,int,bool>? isWater = null,
        Func<int,int,string?>? bridgeAxisAt = null, int maxWaterRun = 0, double turnPenalty = 0, double heuristicWeight = 1)
    {
        int startIndex = start.Y * width + start.X, goalIndex = goal.Y * width + goal.X, startKey = startIndex * 5;
        int[] cameFrom = new int[width * height * 5]; Array.Fill(cameFrom, -1);
        double[] scores = new double[cameFrom.Length]; Array.Fill(scores, double.PositiveInfinity); scores[startKey] = 0;
        var heap = new StableMinHeap<State>(s => s.Score);
        heap.Push(new(startIndex,-1,0,startKey,0,(Math.Abs(goal.X-start.X)+Math.Abs(goal.Y-start.Y))*heuristicWeight));
        Point[] directions = [new(1,0),new(0,1),new(-1,0),new(0,-1)];
        int? goalKey = startIndex == goalIndex ? startKey : null;
        while (heap.Count > 0)
        {
            var current = heap.Pop(); if(current.Cost != scores[current.Key]) continue;
            if(current.Index == goalIndex) { goalKey = current.Key; break; }
            int x = current.Index % width, y = current.Index / width; bool currentWater = isWater?.Invoke(x,y) ?? false;
            for(int direction=0;direction<4;direction++)
            {
                int nx=x+directions[direction].X,ny=y+directions[direction].Y;
                if(nx<0||ny<0||nx>=width||ny>=height) continue;
                bool nextWater=isWater?.Invoke(nx,ny)??false;
                if(currentWater&&current.Direction>=0&&direction!=current.Direction)continue;
                string axis=direction%2==0?"ew":"ns"; string? existing=bridgeAxisAt?.Invoke(nx,ny);
                if(nextWater&&existing!=null&&existing!=axis)continue;
                int waterRun=nextWater?(currentWater?current.WaterRun+1:1):0;
                if(nextWater&&maxWaterRun>0&&waterRun>maxWaterRun)continue;
                int next=ny*width+nx,key=next*5+direction+1;
                double candidate=scores[current.Key]+costAt(nx,ny,x,y)+(current.Direction>=0&&current.Direction!=direction?turnPenalty:0);
                if(candidate>=scores[key])continue;
                scores[key]=candidate;cameFrom[key]=current.Key;
                heap.Push(new(next,direction,waterRun,key,candidate,candidate+(Math.Abs(goal.X-nx)+Math.Abs(goal.Y-ny))*heuristicWeight));
            }
        }
        if(goalKey==null)return [];
        var path=new List<Point>();int cursor=goalKey.Value;
        while(cursor>=0){int cell=cursor/5;path.Add(new(cell%width,cell/width));if(cursor==startKey)break;cursor=cameFrom[cursor];}
        path.Reverse();return path;
    }
}
